cmake: Build `test_bitcoin-qt` executable

hebasto opened the pull request:
A new configuration option is `BUILD_GUI_TESTS`.

---

**Native Windows / MSVC Notes:**
1. vcpkg configures Qt with `-opengl dynamic`, which makes the "minimal" platform plugin unusable due to internal Qt bugs.
2. `ctest` workarounds this issue by setting the environment variable `QT_QPA_PLATFORM=windows`.
3. This approach works nice in the CI for both `x64-windows` and `x64-windows-static` triplets. Please note that `test_bitcoin-qt.exe` fails to run in the CI environment on the master branch.
4. One who is willing to debug Qt with the "minimal" platform plugin could run the `test_bitcoin-qt.exe` directly (not driven by `ctest`).

theuni reviewed (commented):
I noticed while poking around that cmake adds:
`CMakeFiles/test_bitcoin-qt.dir/__/__/__/depends/x86_64-pc-linux-gnu/lib/cmake/Qt5Gui/Qt5Gui_QXcbIntegrationPlugin_Import.cpp.o`

The contents of `Qt5Gui_QXcbIntegrationPlugin_Import.cpp` are:
```c++
#include <QtPlugin>
Q_IMPORT_PLUGIN(QXcbIntegrationPlugin)
```
So I think we're doing that twice?

It seems we should either turn that off somehow, or remove the import from our code?

hebasto commented on .github/workflows/cmake.yml:
```diff
@@ -448,7 +448,7 @@ jobs:
       - name: Test Release configuration
         run: |
-          ctest --test-dir build -j $env:NUMBER_OF_PROCESSORS -C Release
+          ctest --test-dir build -j $env:NUMBER_OF_PROCESSORS -C Release --exclude-regex test_bitcoin-qt
```
A "TODO" comment with links to the issues has been added.

hebasto commented:
> I noticed while poking around that cmake adds: `CMakeFiles/test_bitcoin-qt.dir/__/__/__/depends/x86_64-pc-linux-gnu/lib/cmake/Qt5Gui/Qt5Gui_QXcbIntegrationPlugin_Import.cpp.o`
> 
> The contents of `Qt5Gui_QXcbIntegrationPlugin_Import.cpp` are:
> 
> ```c++
> #include <QtPlugin>
> Q_IMPORT_PLUGIN(QXcbIntegrationPlugin)
> ```
> 
> So I think we're doing that twice?
> 
> It seems we should either turn that off somehow, or remove the import from our code?

Nice catch! It happens on the staging branch as well. Investigating...

hebasto commented:
> I noticed while poking around that cmake adds: `CMakeFiles/test_bitcoin-qt.dir/__/__/__/depends/x86_64-pc-linux-gnu/lib/cmake/Qt5Gui/Qt5Gui_QXcbIntegrationPlugin_Import.cpp.o`
> 
> The contents of `Qt5Gui_QXcbIntegrationPlugin_Import.cpp` are:
> 
> ```c++
> #include <QtPlugin>
> Q_IMPORT_PLUGIN(QXcbIntegrationPlugin)
> ```
> 
> So I think we're doing that twice?
> 
> It seems we should either turn that off somehow, or remove the import from our code?

@theuni 

Thank you for this insight! Please consider https://github.com/hebasto/bitcoin/pull/103.

This PR has been rebased on https://github.com/hebasto/bitcoin/pull/103 and drafted for now.

hebasto commented:
On Windows, the statically linked `test_bitcoin.exe` crashes with the minimal platform plugin, which is used by default.

Specifically, `WalletTests` and `AddressBookTests` fail due to an unhandled exception in the `QMessageBox` class.

This behavior differs from our custom static Qt build in the master branch, which does not crash :)

I believe that the root of the difference in behavio lies in the following Qt configuration options:
- master branch: `-opengl desktop`
- [vcpkg](https://github.com/microsoft/vcpkg/blob/215a2535590f1f63788ac9bd2ed58ad15e6afdff/ports/qt5-base/portfile.cmake#L265): `-opengl dynamic` (a [recommended](https://doc.qt.io/qt-5/windows-requirements.html#dynamically-loading-graphics-drivers) choice)

There are two obvious workarounds for this unfortunate issue:
1. Keep using our custom static Qt build for statically linked executables.
2. Disable `WalletTests` and `AddressBookTests` for the minimal platform plugin as we do on macOS.

Thought?

cc @sipsorcery

sipsorcery commented:
+1 to disable the tests for the dynamically linked version.

Reasoning being that the Windows build is for "developers only" and tracking down Qt bugs for a non-production build doesn't seem to offer a good effort to reward tradeoff.

pablomartin4btc reviewed (commented):
tACK cb1a44fb4696103f56a46488d505b98b04875ea0

Tested on Ubuntu 22.04.

```
...
Tests:
  test_bitcoin ........................ ON
  test_bitcoin-qt ..................... ON
...

```

`./build/src/qt/test/test_bitcoin-qt`

```
Warning: Ignoring XDG_SESSION_TYPE=wayland on Gnome. Use QT_QPA_PLATFORM=wayland to run on Wayland anyway.
********* Start testing of AppTests *********
Config: Using QtTest library 5.15.3, Qt 5.15.3 (x86_64-little_endian-lp64 shared (dynamic) release build; by GCC 11.3.0), ubuntu 22.04
...

All tests passed.

~InitExecutor : Stopping thread
~InitExecutor : Stopped thread

```

hebasto commented:
https://github.com/hebasto/bitcoin/pull/101#discussion_r1501023827:

> Why? Because it requires a gui? Could you please add a comment?

> A "TODO" comment with links to the issues has been added.

The mentioned issues have been fixed in https://github.com/bitcoin-core/gui/pull/803. Please review it as well :)

hebasto commented:
Reworked. Rebased. Ready for review.

The PR description has been updated with Windows/MSVC specific notes.

Friendly ping @TheCharlatan @m3dwards @pablomartin4btc @vasild; and @sipsorcery (as a Windows expert) :)

sipsorcery commented:
With a dynamic linked build on Windows I get:

```
PS C:\dev\github\hebasto-bitcoin\build> ctest -j $env:NUMBER_OF_PROCESSORS -C Release
...
...
99% tests passed, 1 tests failed out of 129

Total Test time (real) = 504.59 sec

The following tests FAILED:
         35 - fs_tests:fs_tests.cpp (Failed)
Errors while running CTest
Output from these tests are in: C:/dev/github/hebasto-bitcoin/build/Testing/Temporary/LastTest.log
Use "--rerun-failed --output-on-failure" to re-run the failed cases verbosely.
```

hebasto commented:
> With a dynamic linked build on Windows I get:
> 
> ```
> PS C:\dev\github\hebasto-bitcoin\build> ctest -j $env:NUMBER_OF_PROCESSORS -C Release
> ...
> ...
> 99% tests passed, 1 tests failed out of 129
> 
> Total Test time (real) = 504.59 sec
> 
> The following tests FAILED:
>          35 - fs_tests:fs_tests.cpp (Failed)
> Errors while running CTest
> Output from these tests are in: C:/dev/github/hebasto-bitcoin/build/Testing/Temporary/LastTest.log
> Use "--rerun-failed --output-on-failure" to re-run the failed cases verbosely.
> ```

Yes. That happens when a user has no [privilege](https://learn.microsoft.com/en-us/previous-versions/windows/it-pro/windows-10/security/threat-protection/security-policy-settings/create-symbolic-links) to create symbolic links. This issue is not specific to this PR change. However, it should be fixed.

sipsorcery commented:
tACK 4da7634698e49486e61b289a27a8c7fc5891bed9.

pablomartin4btc reviewed (approved):
re tACK e4b4c98d43b68b2de3864a436bb343eac216615e

On Ubuntu 22.04.

```
./build/src/qt/test/test_bitcoin-qt

...

All tests passed.
...

```

```
ctest -j $(nproc)

...

100% tests passed, 0 tests failed out of 128

Total Test time (real) =  22.90 sec
```

pablomartin4btc reviewed (approved):
re tACK e4b4c98d43b68b2de3864a436bb343eac216615e

Tested cross-building for Windows on WSL Ubuntu 22.04.

`All tests passed.` (running  `test_bitcoin-qt.exe` successfully in both WSL and Windows 11 Pro)
